File: FSUIKit/Classes/FSUIKit.m
//
//  FSUIKit.m
//  FBRetainCycleDetector
//
//  Created by Fudongdong on 2018/4/3.
//

#import "FSUIKit.h"
#import "FSCalculator.h"

static CGRect oldframe;

@implementation FSUIKit

+ (UIAlertController *)alert:(UIAlertControllerStyle)style controller:(UIViewController *)pController title:(NSString *)title message:(NSString *)message actionTitles:(NSArray<NSString *> *)titles styles:(NSArray<NSNumber *> *)styles handler:(void (^)(UIAlertAction *action))handler cancelTitle:(NSString *)cancelTitle cancel:(void (^)(UIAlertAction *action))cancel completion:(void (^)(void))completion {
#if DEBUG
    NSMutableArray *clears = [[NSMutableArray alloc] init];
    for (int x = 0; x < titles.count; x ++) {
        if (![clears containsObject:titles[x]]) {
            [clears addObject:titles[x]];
        }
    }
    NSAssert(clears.count == titles.count, @"有重复title");
#endif
    UIAlertController *controller = [self alertControllerWithStyle:style title:title message:message actionTitles:titles styles:styles handler:handler cancelTitle:cancelTitle cancel:cancel];
    [pController presentViewController: controller animated:YES completion: completion];
    return controller;
}

+ (UIAlertController *)alert:(UIAlertControllerStyle)style controller:(UIViewController *)pController title:(NSString *)title message:(NSString *)message actionTitles:(NSArray<NSString *> *)titles styles:(NSArray<NSNumber *> *)styles handler:(void (^)(FSAlertAction *action))handler {
    UIAlertController *controller = [self alertControllerWithStyle:style title:title message:message actionTitles:titles styles:styles handler:handler cancelTitle:@"取消" cancel:nil];
    [pController presentViewController:controller animated:YES completion:nil];
    return controller;
}

+ (UIAlertController *)alertControllerWithStyle:(UIAlertControllerStyle)style title:(NSString *)title message:(NSString *)message actionTitles:(NSArray<NSString *> *)titles styles:(NSArray<NSNumber *> *)styles handler:(void (^)(FSAlertAction *action))handler cancelTitle:(NSString *)cancelTitle cancel:(void (^)(FSAlertAction *action))cancel{
    if ([UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPad) {
        style = UIAlertControllerStyleAlert;
    }
    
    NSInteger count = MIN(titles.count, styles.count);
    UIAlertController *controller = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:style];
    for (int x = 0; x < count; x ++) {
        FSAlertAction *action = [FSAlertAction actionWithTitle:titles[x] style:[styles[x] integerValue] handler:^(UIAlertAction * _Nonnull action) {
            if (handler) {
                handler((FSAlertAction *)action);
            }
        }];
        action.theTag = x;
        [controller addAction:action];
    }
    if ([cancelTitle isKindOfClass:NSString.class] && cancelTitle.length) {
        FSAlertAction *archiveAction = [FSAlertAction actionWithTitle:cancelTitle style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
            if (cancel) {
                cancel((FSAlertAction *)action);
            }
        }];
        [controller addAction:archiveAction];
    }
    return controller;
}

+ (UIAlertController *)alertInput:(NSInteger)number controller:(UIViewController *)controller title:(NSString *)title message:(NSString *)message buttons:(NSInteger)buttons buttonConfig:(void (^)(FSAlertActionData *data))buttonConfig textFieldConifg:(void (^)(UITextField *textField))textFieldConfig completion:(void (^)(UIAlertController *alert))completion {
    if (number < 1) {
        return nil;
    }
    
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle: title message: message preferredStyle: UIAlertControllerStyleAlert];
    for (int x = 0; x < number; x ++) {
        [alertController addTextFieldWithConfigurationHandler:^(UITextField *textField) {
            if (textFieldConfig) {
                textField.tag = x;
                textFieldConfig(textField);
            }
        }];
    }
    
    __weak typeof(alertController)wAlertController = alertController;
    for (int x = 0; x < buttons; x ++) {
        FSAlertActionData *data = [[FSAlertActionData alloc] init];
        data.index = x;
        data.style = UIAlertActionStyleDefault;
        
        if (buttonConfig) {
            buttonConfig(data);
        }
        
        FSAlertAction *action = [FSAlertAction actionWithTitle: data.title style: data.style handler: ^(UIAlertAction * _Nonnull action) {
            FSAlertAction *act = (FSAlertAction *)action;
            if (act.data.click) {
                act.data.click(wAlertController, action);
            }
        }];
        action.data = data;
        [alertController addAction: action];
    }
    
    [controller presentViewController: alertController animated: YES completion:^{
        if (completion) {
            completion(alertController);
        }
    }];
    return alertController;
}

+ (void)showAlertWithMessage:(NSString *)message controller:(UIViewController *)controller{
    [self showAlertWithTitle:@"温馨提示" message:message ok:@"确定" controller:controller handler:nil];
}

+ (void)showAlertWithMessage:(NSString *)message controller:(UIViewController *)controller handler:(void (^)(UIAlertAction *action))handler{
    [self showAlertWithTitle:@"温馨提示" message:message ok:@"确定" controller:controller handler:handler];
}

+ (void)showAlertWithTitle:(NSString *)title message:(NSString *)message ok:(NSString *)ok controller:(UIViewController *)pController handler:(void (^)(UIAlertAction *action))handler{
    UIAlertController *controller = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *action = [UIAlertAction actionWithTitle:ok style:UIAlertActionStyleDefault handler:handler];
    [controller addAction:action];
    [pController presentViewController:controller animated:YES completion:nil];
}

+ (UIImage *)captureScrollView:(UIScrollView *)scrollView{
    CGRect frame = scrollView.frame;
    //设置控件显示的区域大小     key:显示
    scrollView.frame = CGRectMake(0, scrollView.frame.origin.y, scrollView.contentSize.width, scrollView.contentSize.height);
    
    //设置截屏大小(截屏区域的大小必须要跟视图控件的大小一样)
    UIGraphicsBeginImageContextWithOptions(scrollView.frame.size, YES, 0.0);
    [[scrollView layer] renderInContext:UIGraphicsGetCurrentContext()];
    UIImage *viewImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    scrollView.frame = frame;
    return viewImage;
}

+ (void)captureScrollView:(UIScrollView *)scrollView finished:(void(^)(UIImage *image))completion{
    if (!completion) {
        return;
    }
    
    UIImage * tableViewScreenshot = [FSImage imageForUIView:scrollView];
    completion(tableViewScreenshot);
}

//截取view生成一张图片
+ (UIImage *)shotWithView:(UIView *)view{
    UIGraphicsBeginImageContext(view.bounds.size);
    [view.layer renderInContext:UIGraphicsGetCurrentContext()];
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return image;
}

//Avilable in iOS 8.0 and later
+ (UIVisualEffectView *)effectViewWithFrame:(CGRect)frame{
    UIBlurEffect *effect = [UIBlurEffect effectWithStyle:UIBlurEffectStyleLight];
    UIVisualEffectView *effectView = [[UIVisualEffectView alloc] initWithEffect:effect];
    effectView.frame = frame;
    return effectView;
}

//截取view中某个区域生成一张图片
+ (UIImage *)shotWithView:(UIView *)view scope:(CGRect)scope{
    CGImageRef imageRef = CGImageCreateWithImageInRect([self shotWithView:view].CGImage, scope);
    UIGraphicsBeginImageContext(scope.size);
    CGContextRef context = UIGraphicsGetCurrentContext();
    CGRect rect = CGRectMake(0, 0, scope.size.width, scope.size.height);
    CGContextTranslateCTM(context, 0, rect.size.height);//下移
    CGContextScaleCTM(context, 1.0f, -1.0f);//上翻
    CGContextDrawImage(context, rect, imageRef);
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    CGImageRelease(imageRef);
    return image;
}

+ (UIImage*)circleImage:(UIImage*)image withParam:(CGFloat)inset{
    UIGraphicsBeginImageContext(image.size);
    CGContextRef context =UIGraphicsGetCurrentContext();
    
    //圆的边框宽度为2，颜色为红色
    CGContextSetLineWidth(context,2);
    CGContextSetStrokeColorWithColor(context, [UIColor clearColor].CGColor);
    CGRect rect = CGRectMake(inset, inset, image.size.width - inset *2.0f, image.size.height - inset *2.0f);
    CGContextAddEllipseInRect(context, rect);
    CGContextClip(context);
    
    //在圆区域内画出image原图
    [image drawInRect:rect];
    CGContextAddEllipseInRect(context, rect);
    CGContextStrokePath(context);
    
    //生成新的image
    UIImage *newimg = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return newimg;
}

+ (UIImage *)QRImageFromString:(NSString *)sourceString{
    if (sourceString == nil) {
        sourceString = @"";
    }
    CIFilter *filter = [CIFilter filterWithName:@"CIQRCodeGenerator"];
    [filter setDefaults];
    NSData *data = [sourceString dataUsingEncoding:NSUTF8StringEncoding];
    [filter setValue:data forKey:@"inputMessage"];
    CIImage *ciImage = [filter outputImage];
    //UIImage *unsharpImage = [UIImage imageWithCIImage:ciImage];//不清晰
    UIImage *image = [self createNonInterpolatedUIImageFormCIImage:ciImage withSize:960];
    return image;
}

+ (UIImage *)createNonInterpolatedUIImageFormCIImage:(CIImage *)image withSize:(CGFloat)size{
    CGRect extent = CGRectIntegral(image.extent);
    CGFloat scale = MIN(size/CGRectGetWidth(extent), size/CGRectGetHeight(extent));
    size_t width = CGRectGetWidth(extent) * scale;
    size_t height = CGRectGetHeight(extent) * scale;
    CGColorSpaceRef cs = CGColorSpaceCreateDeviceGray();
    CGContextRef bitmapRef = CGBitmapContextCreate(nil, width, height, 8, 0, cs, (CGBitmapInfo)kCGImageAlphaNone);
    CIContext *context = [CIContext contextWithOptions:nil];
    CGImageRef bitmapImage = [context createCGImage:image fromRect:extent];
    CGContextSetInterpolationQuality(bitmapRef, kCGInterpolationNone);
    CGContextScaleCTM(bitmapRef, scale, scale);
    CGContextDrawImage(bitmapRef, extent, bitmapImage);
    CGImageRef scaledImage = CGBitmapContextCreateImage(bitmapRef);
    CGContextRelease(bitmapRef);
    CGImageRelease(bitmapImage);
    CGColorSpaceRelease(cs);
    UIImage *result = [UIImage imageWithCGImage:scaledImage];
    CGImageRelease(scaledImage);
    return result;
}

/*
 ** lineFrame:     虚线的 frame
 ** length:        虚线中短线的宽度
 ** spacing:       虚线中短线之间的间距
 ** color:         虚线中短线的颜色
 */
+ (UIView *)createDashedLineWithFrame:(CGRect)lineFrame
                           lineLength:(int)length
                          lineSpacing:(int)spacing
                            lineColor:(UIColor *)color{
    UIView *dashedLine = [[UIView alloc] initWithFrame:lineFrame];
    dashedLine.backgroundColor = [UIColor clearColor];
    CAShapeLayer *shapeLayer = [CAShapeLayer layer];
    [shapeLayer setBounds:dashedLine.bounds];
    [shapeLayer setPosition:CGPointMake(CGRectGetWidth(dashedLine.frame) / 2, CGRectGetHeight(dashedLine.frame))];
    [shapeLayer setFillColor:[UIColor clearColor].CGColor];
    [shapeLayer setStrokeColor:color.CGColor];
    [shapeLayer setLineWidth:CGRectGetHeight(dashedLine.frame)];
    [shapeLayer setLineJoin:kCALineJoinRound];
    [shapeLayer setLineDashPattern:[NSArray arrayWithObjects:[NSNumber numberWithInt:length], [NSNumber numberWithInt:spacing], nil]];
    CGMutablePathRef path = CGPathCreateMutable();
    CGPathMoveToPoint(path, NULL, 0, 0);
    CGPathAddLineToPoint(path, NULL, CGRectGetWidth(dashedLine.frame), 0);
    [shapeLayer setPath:path];
    CGPathRelease(path);
    [dashedLine.layer addSublayer:shapeLayer];
    return dashedLine;
}

+ (CGFloat)scrollViewPage:(UIScrollView *)scrollView {
    CGFloat pageWidth = scrollView.frame.size.width;
    if (pageWidth < 0.01) {
        return 0;
    }
    CGFloat page = floor((scrollView.contentOffset.x - pageWidth / 2) / pageWidth) + 1;
    int currentPage = page;
    return currentPage;
}

@end

File: FSUIKit/Classes/Swift/FSUIKit.swift
// FSUIKitS.swift
// Translated from FSUIKit.h/m

import UIKit

@objcMembers
public class FSUIKit: NSObject {
    
    public static func alert(style: UIAlertController.Style, controller: UIViewController, title: String, message: String, actionTitles: [String], styles: [UIAlertAction.Style], handler: ((UIAlertAction) -> Void)?, cancelTitle: String, cancel: ((UIAlertAction) -> Void)?, completion: (() -> Void)?) -> UIAlertController {
        
        let alertController = self.alertControllerWithStyle(style, title: title, message: message, actionTitles: actionTitles, styles: styles, handler: handler, cancelTitle: cancelTitle, cancel: cancel);
        controller.present(alertController, animated: true, completion: completion)
        return alertController
    }

    static func alert(style: UIAlertController.Style, controller: UIViewController, title: String, message: String, actionTitles: [String], styles: [UIAlertAction.Style], handler: ((FSAlertAction) -> Void)?) -> UIAlertController {
        
        let ac = self.alertControllerWithStyle(style, title: title, message: message, actionTitles: actionTitles, styles: styles, handler: handler, cancelTitle: "取消", cancel: nil)
        controller.present(ac, animated: true, completion: nil)
        return ac
    }
    
    static func alertControllerWithStyle(_ style1: UIAlertController.Style, title:String, message: String, actionTitles: [String], styles: [UIAlertAction.Style], handler: ( (FSAlertAction) -> Void)?, cancelTitle: String, cancel: ((FSAlertAction) -> Void)?) -> UIAlertController {
        
        var style = style1
        if UIDevice.current.userInterfaceIdiom == .pad {
            style = UIAlertController.Style.alert
        }
        
        let count = min(actionTitles.count, styles.count)
        let controller = UIAlertController(title: title, message: message, preferredStyle: style)
        
        for i in 0..<count {
            let action = FSAlertAction(title: actionTitles[i], style: styles[i], handler: { act in
                if handler != nil {
                    handler!(act as! FSAlertAction)
                }
            })
            action.theTag = i
            controller.addAction(action)
        }
        
        let archiveAction = FSAlertAction(title: cancelTitle, style: .cancel) { act in
            if cancel != nil {
                cancel!(act as! FSAlertAction)
            }
        }
        
        controller.addAction(archiveAction)
        
        return controller
    }
    
    static func alertInput(number: Int, controller: UIViewController, title: String, message: String, buttons: Int, buttonConfig: (FSAlertActionData) -> Void, textFieldConfig: ((UITextField) -> Void)?, completion: ((UIAlertController) -> Void)?) -> UIAlertController {
        
        if number < 1 {
            return UIAlertController()
        }
        
        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
        
        for i in 0..<number {
            alert.addTextField { tf in
                if textFieldConfig != nil {
                    tf.tag = i
                    textFieldConfig!(tf)
                }
            }
        }
        
        for i in 0..<buttons {
            let data = FSAlertActionData()
            data.index = i
            data.style = .default

            buttonConfig(data)
            
            let action = FSAlertAction(title: data.title, style: data.style) { [weak alert] m in
                guard let alert else { return }

                let act: FSAlertAction = m as! FSAlertAction
                if act.data.click != nil {
                    act.data.click!(alert, act)
                }
            }
            
            action.data = data
            alert.addAction(action)
        }
        
        controller.present(alert, animated: true) {
            if completion != nil {
                completion!(alert)
            }
        }
        
        return alert
    }
    
    public static func showAlertWithMessage(message: String, controller: UIViewController) {
        self.showAlertWithTitle(title: "温馨提示", message: message, ok: "确定", controller: controller, handler: nil)
    }
    
    public static func showAlertWithMessage(message: String, controller: UIViewController, handler: ((UIAlertAction) -> Void)?) {
        self.showAlertWithTitle(title: "温馨提示", message: message, ok: "确定", controller: controller, handler: handler)
    }
    
    public static func showAlertWithTitle(title: String, message: String, ok: String, controller: UIViewController, handler: ((UIAlertAction) -> Void)?) {
        let c = UIAlertController(title: title, message: message, preferredStyle: .alert)
        let action = UIAlertAction(title: ok, style: .default, handler: handler)
        c.addAction(action)
        controller.present(c, animated: true, completion: nil)
    }
    
    @available(iOS 8.0, *)
    public static func effectView(withFrame frame: CGRect) -> UIVisualEffectView {
        let effect = UIBlurEffect(style: .light)
        let effectView = UIVisualEffectView(effect: effect)
        effectView.frame = frame
        return effectView
    }
    
    public static func shot(with view: UIView, scope: CGRect) -> UIImage? {
        guard let fullImage = shot(with: view),
              let imageRef = fullImage.cgImage?.cropping(to: scope) else { return nil }
        
        UIGraphicsBeginImageContext(scope.size)
        guard let context = UIGraphicsGetCurrentContext() else { return nil }
        
        let rect = CGRect(x: 0, y: 0, width: scope.width, height: scope.height)
        context.translateBy(x: 0, y: rect.height)
        context.scaleBy(x: 1.0, y: -1.0)
        context.draw(imageRef, in: rect)
        
        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return image
    }
    
    public static func shot(with view: UIView) -> UIImage? {
        UIGraphicsBeginImageContext(view.bounds.size)
        view.layer.render(in: UIGraphicsGetCurrentContext()!)
        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return image
    }
    
    public static func captureScrollView(_ scrollView: UIScrollView, finished completion: @escaping (UIImage?) -> Void) {
        let image = FSImageS.image(for: scrollView)
        completion(image)
    }
    
    public static func createDashedLine(withFrame lineFrame: CGRect, lineLength length: Int, lineSpacing spacing: Int, lineColor color: UIColor) -> UIView {
        let dashedLine = UIView(frame: lineFrame)
        dashedLine.backgroundColor = .clear
        
        let shapeLayer = CAShapeLayer()
        shapeLayer.bounds = dashedLine.bounds
        shapeLayer.position = CGPoint(x: dashedLine.frame.width / 2, y: dashedLine.frame.height)
        shapeLayer.fillColor = UIColor.clear.cgColor
        shapeLayer.strokeColor = color.cgColor
        shapeLayer.lineWidth = dashedLine.frame.height
        shapeLayer.lineJoin = .round
        shapeLayer.lineDashPattern = [NSNumber(value: length), NSNumber(value: spacing)]
        
        let path = CGMutablePath()
        path.move(to: CGPoint(x: 0, y: 0))
        path.addLine(to: CGPoint(x: dashedLine.frame.width, y: 0))
        shapeLayer.path = path
        
        dashedLine.layer.addSublayer(shapeLayer)
        return dashedLine
    }
    
    public static func qrImage(from string: String?) -> UIImage? {
        let sourceString = string ?? ""
        guard let filter = CIFilter(name: "CIQRCodeGenerator") else { return nil }
        filter.setDefaults()
        
        let data = sourceString.data(using: .utf8)
        filter.setValue(data, forKey: "inputMessage")
        
        guard let ciImage = filter.outputImage else { return nil }
        return createNonInterpolatedUIImage(from: ciImage, withSize: 960)
    }
    
    private static func createNonInterpolatedUIImage(from image: CIImage, withSize size: CGFloat) -> UIImage? {
        let extent = image.extent.integral
        let scale = min(size / extent.width, size / extent.height)
        let width = Int(extent.width * scale)
        let height = Int(extent.height * scale)
        
        guard let cs = CGColorSpace(name: CGColorSpace.linearGray),
              let bitmapRef = CGContext(data: nil, width: width, height: height, bitsPerComponent: 8, bytesPerRow: 0, space: cs, bitmapInfo: CGImageAlphaInfo.none.rawValue) else { return nil }
        
        let context = CIContext(options: nil)
        guard let bitmapImage = context.createCGImage(image, from: extent) else { return nil }
        
        bitmapRef.interpolationQuality = .none
        bitmapRef.scaleBy(x: scale, y: scale)
        bitmapRef.draw(bitmapImage, in: extent)
        
        guard let scaledImage = bitmapRef.makeImage() else { return nil }
        return UIImage(cgImage: scaledImage)
    }
    
    public static func circleImage(_ image: UIImage, withParam inset: CGFloat) -> UIImage? {
        UIGraphicsBeginImageContext(image.size)
        guard let context = UIGraphicsGetCurrentContext() else { return nil }
        
        context.setLineWidth(2)
        context.setStrokeColor(UIColor.clear.cgColor)
        
        let rect = CGRect(x: inset, y: inset, width: image.size.width - inset * 2.0, height: image.size.height - inset * 2.0)
        context.addEllipse(in: rect)
        context.clip()
        
        image.draw(in: rect)
        context.addEllipse(in: rect)
        context.strokePath()
        
        let newImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return newImage
    }
    
    public static func scrollViewPage(_ scrollView: UIScrollView) -> CGFloat {
        let pageWidth = scrollView.frame.width
        if pageWidth < 0.01 { return 0 }
        let page = floor((scrollView.contentOffset.x - pageWidth / 2) / pageWidth) + 1
        return page
    }
}
